### PR TITLE
Df asm observe

### PR DESCRIPTION
**Pull request by dflanderka**
- implementation of new assembly algorithm in observe output

### CODE REVIEW

File: src/fields/equation_output.hh [hunk 1]
```diff
@@ -93,9 +97,29 @@ public:
      */
     void output(TimeStep step);
 
+    /// Getter for set of observe fields.
+    inline const std::unordered_set<string> &observe_fields() const {
```
Comment (jbrezmorf) [inline src/fields/equation_output.hh:101]: We should not pass the whole EquationOutput to the observe assembly and implement those getters, rather pass necessary data through the constructor, possibly wrapping them into an auxiliary struct.

File: src/fields/field.impl.hh [hunk 2]
```diff
@@ -398,8 +398,9 @@ void Field<spacedim, Value>::observe_output(std::shared_ptr<Observe> observe)
 
     if (observe->point_ds()->size() == 0) return;
```
Comment (jbrezmorf) [inline src/fields/field.impl.hh:400]: observe_output would be necessary only for DarcyMH, so we should remove DarcyMH first (in master) and remove this method as well.

File: src/fields/field.hh [hunk 3]
```diff
@@ -424,6 +433,9 @@ protected:
     /// ElementDataCache used during field output, object is shared with OutputTime
     std::shared_ptr<ElementDataCache<typename Value::element_type>> output_data_cache_;
 
+    /// Same as previous but used during observe output, object is shared with Observe
+    std::shared_ptr<ElementDataCache<typename Value::element_type>> observe_data_cache_;
```
Comment (jbrezmorf) [inline src/fields/field.hh:437]: Should not be in Field, rather copy directly to the Observe object.

File: src/fields/equation_output.hh [hunk 4]
```diff
@@ -93,9 +97,29 @@ public:
...
+    /// Getter for OutputTime object.
+    inline Mesh *eq_mesh() {
+        return mesh_;
+    }
+
     /// Selects the error control field out of output field set according to input record.
     typename OutputMeshBase::ErrorControlFieldFunc select_error_control_field();
     
+    /// Getter from patch_point_data.
+    inline PatchPointVec &patch_point_data() {
```
Comment (jbrezmorf) [inline src/fields/equation_output.hh:119]: Can we move this into Observe ?

File: src/fields/equation_output.cc [hunk 5]
```diff
@@ -318,12 +323,20 @@ void EquationOutput::output(TimeStep step)
...
 
     // observe output
-    for(FieldCommon * field : this->field_list) {
-        if ( field->flags().match( FieldFlag::allow_output) ) {
-            if (observe_fields_.find(field->name()) != observe_fields_.end()) {
-                field->observe_output( observe_ptr );
+    {
+        FieldSet used_fields;
+        uint n_fields = 0;
+        for(FieldCommon * f_acc : this->field_list) {
```
Comment (jbrezmorf) [inline src/fields/equation_output.cc:329]: Just form the output field set here and pass it to the assembly together with Observe instance.

File: src/fields/assembly_observe.hh [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+
+private:
+    /// Calls cache_reallocate method on set of used fields
+    inline void reallocate_cache() {
+        multidim_assembly_[1_d]->eq_fields_->cache_reallocate(this->element_cache_map_, multidim_assembly_[1_d]->used_fields_);
+        // DebugOut() << "Order of evaluated fields (" << DimAssembly<1>::name() << "):" << multidim_assembly_[1_d]->eq_fields_->print_dependency();
+    }
+
+    std::shared_ptr<EvalPoints> eval_points_;                     ///< EvalPoints object
```
Comment (jbrezmorf) [inline src/fields/assembly_observe.hh:126]: Aren't these in GenericAssembly? Why we have to declare them again?

File: src/fields/assembly_observe.hh [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+struct PatchPointData {
+    /// Default constructor
+    PatchPointData() {}
+
+    /// Constructor with data mebers initialization
+    PatchPointData(unsigned int elm_idx, arma::vec loc_coords)
+    : elem_idx(elm_idx), local_coords(loc_coords), i_quad(0), i_quad_point(0) {}
+
+    /// Copy constructor
+    PatchPointData(const PatchPointData &other)
```
Comment (jbrezmorf) [inline src/fields/assembly_observe.hh:40]: No constructors necessary for POD struct types.

Constructed simply by:
`PatchPointData x = {elm_idx, loc_coords, 0,0}`

or `emplace_back(idx, loc_coords, 0,0)`

**Review by jbrezmorf — CHANGES_REQUESTED** [review]
Deserves some simplifications.

File: src/fields/assembly_observe.hh (hunk 6 above)
Comment (jbrezmorf) [reply src/fields/assembly_observe.hh:126]: Move to common parent class ?

File: src/fields/assembly_observe.hh [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    : AssemblyBase<dim>(), eq_fields_(eq_fields), eq_data_(eq_data) {
+        this->active_integrals_ = ActiveIntegrals::bulk;
+        offsets_.resize(1.1 * CacheMapElementNumber::get());
+
+        for (auto observe_field : eq_data_->observe_fields()) {
+            auto found_field = eq_fields_->field(observe_field);
+            used_fields_ += *found_field;
+        }
+
+        if (dim==1) {
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/fields/assembly_observe.hh:154]: Create `patch_point_data` in EquationOutput and pass to the Assembly as well.

File: src/fields/equation_output.hh (hunk 4 above)
Comment (jbrezmorf) [reply src/fields/equation_output.hh:119]: Move `patch_point_data_` vector into Observe class.

File: src/fields/field.impl.hh (hunk 2 above)
Comment (dflanderka) [reply src/fields/field.impl.hh:400]: observe_output is not used now, method will be removed, assembly classes used in equations and AssemblyObserveOutput class are independent.

Comment (dflanderka) [reply src/fields/field.impl.hh:400]: observe_output is not used now, method will be removed, assembly classes used in equations and AssemblyObserveOutput class are independent.

**Pull request merged**